doc/source/extended_documentation/regrid/bilinear_land_sea.rst

```diff
@@ -0,0 +1,132 @@
+################################################
+Bilinear regridding with land-sea mask awareness
+################################################
+
+This subpackage provides a fast and unified implementation of bilinear regridding with land-sea mask awareness. It is assumed that the source grid is evenly-spaced in a rectilinear latitude/longitude system. The target grid could be rectilinear or curvilinear. 
+
+********************************
+Calculation steps of Regridding 
+********************************
+ 
+* Given a target point, determine its four source points according to regular grid point ordering.
+
+* Identify surface type of source/target points from land sea mask input files.
+
+* Identify if source points have the same surface type as the target point. 
+  The matched source points are called "True", and unmatched source points are called "False". 
+
+* Calculate weights of four source points with normal bilinear formulation.
+
+* Adjust source points & weights if unmatched source points exist. 
+
+* Apply weights at values of source points for interpolated value in a unified way
+  so that vectorisation of the Numpy computation can be fully achieved.
+  
+*************************************
+The weight calculation and adjustment
+*************************************
+
+::
+                        
+                4   .----------. 3     
+                    |          |       target point x(lat,lon)        
+                lon |    x     |       four source points 1,2,3 and 4 
+                    |          |
+                1   .----------. 2
+                        lat
+ 
+
+1. For the target points with four matched surrounding source points, their source point
+   weights can be  calculated using the Finite-Element based interpolation shape function. 
+
+.. math:: 
+        var &= \sum \limits_{i=1}^4 weight_i \cdot var_i
+      
+        weight_1 &= (lat_2-lat)*(lon_3-lon)/area
+ 
+        weight_2 &= (lat-lat_1)*(lon_3-lon)/area
+        
+        weight_3 &= (lat-lat_1)*(lon-lon_1)/area
+         
+        weight_4 &= (lat_2-lat)*(lon-lon_1)/area
+
+        area &= (lat_2-lat_1)*(lon_3-lon_1)
+
+2.  For the target points with three matched source points plus one unmatached source point, and
+    the target point inside the triangle formed with three matched source points, three matched 
+    source points is used to construct linear field for interpolation. The Finite-Element based
```

Comment:
Have you got a good basic reference for finite-element shape functions and how they are derived so that someone could attempt to understand how these functions have been derived in the future?

Reply:
I prefer not to list any references here, because these formulations have no direct references. The basic idea how to derive the shape formulation is available in any finite element books, if users are interested in deriving it.

Reply:
There isn't anyone else on the team who has a background in finite element methods so I think a link to an example textbook would still be a good idea so someone knows where to start. You can make it clear that you have followed standard methods to derive these formulae and here is a place you can go to look up the basic idea of how to derive these kinds of formulae.

Reply:
Ok. I add a reference.